packages/lib/gpcircuits/src/list-membership.ts

```diff
@@ -0,0 +1,183 @@
+  const unpaddedOutputObject: {
+    tupleIndices: CircuitSignal[][];
+    memberIndices: CircuitSignal[];
+    membershipLists: CircuitSignal[][];
+  } = {
+    tupleIndices: [],
+    memberIndices: [],
+    membershipLists: []
+  };
+
+  for (let i = 0; i < memberIndices.length; i++) {
+    // Process each of the lists
+    const processedList = processSingleList(
+      params,
+      firstTupleIndex,
+      memberIndices[i],
+      lists[i]
+    );
+
+    // Push tuple indices as bigints.
+    processedList.tupleIndices.forEach((indexTuple: number[]) =>
+      unpaddedOutputObject.tupleIndices.push(indexTuple.map(BigInt))
+    );
+    // Push member indices as bigints.
+    unpaddedOutputObject.memberIndices.push(BigInt(processedList.memberIndex));
+
+    // Push the hashed membership list.
+    unpaddedOutputObject.membershipLists.push(processedList.membershipList);
+
+    // Increment the first tuple index.
+    firstTupleIndex += processedList.tupleIndices.length;
+  }
+
+  // Pad and return
+  return {
+    tupleIndices: padArray(
+      unpaddedOutputObject.tupleIndices,
+      params.maxTuples,
+      // Pad with tuples of 0.
+      extendedSignalArray([], params.tupleArity)
+    ),
+    memberIndex: extendedSignalArray(
+      unpaddedOutputObject.memberIndices,
+      params.maxLists,
+      // Pad with -1 (mod p).
+      BABY_JUB_NEGATIVE_ONE
+    ),
+    membershipList: padArray(
+      unpaddedOutputObject.membershipLists,
+      params.maxLists,
+      // Pad with lists of zeroes.
+      extendedSignalArray([], params.maxListEntries)
+    )
+  };
+}
+
+/**
+ * Processes a single membership list together with the (multi-)index
+ * of the entry value (or entry value tuple) that ought to
+ * be a member of this list. This is done by means of appropriate
+ * applications of {@link hashTuple} and {@link computeTupleIndices}.
+ * If no tuples are involved, then `memberIndex` is a singleton and
+ * `list` a list of singletons.
+ *
+ * @param params parameters of the ProtoPODGPC the list is processed for
+ * @param firstTupleIndex the index of the first output tuple in the combined array of
+ * entry and tuple value hashes.
+ * @param memberIndex the index of the entry value (or tuple) which is a member of the list
+ * @param list the membership list of tuples
+ * @returns list of tuple indices of arity `params.tupleArity` representing the input tuple,
+ * number representing the index of the entry value (tuple) which is a member of
+ * the list, and the (unpadded) membership list in hashed form.
+ * @throws RangeError if any of the indices are out of bounds with respect
+ * to the given parameters.
+ * @throws TypeError if the list and index arrays are malformed.
+ */
+export function processSingleList(
+  params: ProtoPODGPCCircuitParams,
+  firstTupleIndex: number,
+  memberIndex: number[],
+  list: PODValue[][]
+): { tupleIndices: number[][]; memberIndex: number; membershipList: bigint[] } {
```

Comment:
I was thinking when the time comes this could be chained with whatever else requires tuples provided the `firstTupleIndex` parameter is appropriately set  on each call (and all `tupleIndices` are concatenated in the right order). `processLists` will have to be modified for this.

Reply:
That's clever.  Maybe calling it firstAvailableTupleIndex would help clarify that purpose?

Something you'll have to figure out as you add this in to the @pcd/gpc layer is how much the definition of these tuples is explicit (i.e. there's a tuple config object) or implicit (i.e. the list config contains an input specification which is a tuple).  There's a lot of complexity here, but it might mean that the tuple definitions can be entirely implicit.